client/shared/src/test/scala/org/http4s/client/middleware/LoggerSuite.scala

```diff
@@ -65,6 +67,31 @@ class LoggerSuite extends Http4sSuite {
     res.assertEquals(expectedBody)
   }
 
+  private val configurableRequestLoggerClient =
+    (logBody: Boolean, logAction: Option[String => IO[Unit]]) =>
+      RequestLogger.apply[IO](true, logBody, logAction = logAction)(
+        Client.fromHttpApp(testApp)
+      )
+
+  test("RequestLogger should log a Get for all values of logBody") {
+    forAllF { (logBody: Boolean) =>
+      val req = Request[IO](uri = uri"/request")
+      val message = "GET /request"
+      val logAction = (q: Queue[IO, String]) =>
+        (actualMessage: String) =>
+          q.tryTake.map(expectedSubstring =>
+            assert(expectedSubstring.exists(actualMessage.contains), "unexpected message logged")
+          )
+
+      for {
+        logger <- Queue.unbounded[IO, String]
+        _ <- logger.offer(message)
+        _ <- configurableRequestLoggerClient(logBody, Some(logAction(logger))).successful(req)
+        remaining <- logger.tryTake
+      } yield assert(remaining.isEmpty, "logAction not called")
```

Comment:
It's late so I'm not thinking about it too hard, but honestly I'm pretty confused by this test 😅 why can't we do something stupid simple, like complete a `Deferred` if it logs or not?

Reply:
In this test, the `Queue` is "primed" with the expected parameters that eventually get passed to `logAction: Option(String => F[Unit]`. The `logAction` itself is asserting that the parameters are as expected, with the final assertion checking that the primed queue has been exhausted. 
I use this pattern frequently, but realize it may be a bit inside out. It's less helpful when priming for only a single expected parameter, I will re-work a bit to use `Deferred` for clarity.

Reply:
Please let me know if this is clearer: 5ecde58bb0bff8cab60f55292869e482346e66c0

Reply:
Thanks for explaining! I suspected it was something like that. But I do find the rewrite much easier to wrap my head around :)